projects/batfish/src/main/java/org/batfish/job/BatfishJobExecutor.java

```diff
@@ -41,105 +50,143 @@ public BatfishJobExecutor(
     }
   }
+
+  ExecutorService getExecutorService() {
+    ExecutorService pool;
+    if (!_settings.getSequential()) {
+      int maxConcurrentThreads = Runtime.getRuntime().availableProcessors();
+      int numConcurrentThreads = Math.min(maxConcurrentThreads, _settings.getJobs());
+      pool = Executors.newFixedThreadPool(numConcurrentThreads);
+    } else {
+      pool = Executors.newSingleThreadExecutor();
+    }
+    return pool;
+  }
+
+  /**
+   * Handles the Job result appropriately in case of success/error, and updates logs
+   *
+   * @param result Result of Job
+   * @param output Output comprising results of all jobs
+   * @param answerElement Answer Element for the execution of jobs
+   * @param failureCauses List of failure causes for all jobs
+   * @return True if there is a processing error else False
+   */
+  boolean handleJobResult(
+      JobResultT result,
+      OutputT output,
+      AnswerElementT answerElement,
+      List<BatfishException> failureCauses) {
+    boolean processingError = false;
+    String time = CommonUtil.getTime(result.getElapsedTime());
+    Throwable failureCause = result.getFailureCause();
+    if (failureCause == null) {
+      result.applyTo(output, _logger, answerElement);
+      _logger.infof(
+          "Job terminated successfully with result: %s after elapsed time: %s - %d/%d "
+              + "(%.1f%%) complete\n",
+          result.toString(), time, _finishedJobs, _totalJobs, _finishedPercent);
+    } else {
+      String failureMessage =
+          String.format(
+              "Failure running job after elapsed time: %s\n-----"
+                  + "BEGIN JOB LOG-----\n%s\n-----END JOB LOG-----",
+              time,
+              result.getHistory().toString(BatfishLogger.getLogLevel(_settings.getLogLevel())));
+      BatfishException bfc = new BatfishException(failureMessage, failureCause);
+      if (_settings.getExitOnFirstError()) {
+        result.appendHistory(_logger);
+        throw bfc;
+      } else {
+        processingError = true;
+        result.appendHistory(_logger);
+        _logger.error(failureMessage + ":\n\t" + ExceptionUtils.getStackTrace(failureCause));
+        failureCauses.add(bfc);
+        if (!_haltOnProcessingError) {
+          result.applyTo(output, _logger, answerElement);
+        }
+      }
+    }
+    return processingError;
+  }
+
+  void handleProcessingError(List<JobT> jobs, List<BatfishException> failureCauses) {
+    int numJobs = jobs.size();
+    int numFailed = numJobs - failureCauses.size();
+    int numSucceeded = numJobs - numFailed;
+    if (!_logger.isActive(BatfishLogger.LEVEL_INFO)) {
+      _logger.infof("%d jobs succeeded; %d jobs failed\n", numSucceeded, numFailed);
+    }
+    if (_haltOnProcessingError) {
+      throw new CompositeBatfishException(
+          new BatfishException("Fatal exception due to failure of at least one job"),
+          failureCauses);
+    }
+  }
+
+  void initializeJobsStats(List<JobT> jobs) {
```

Comment:
I don't quite understand why `BatfishJobExecutor` is a seemingly reusable class with state that is reinitialized per job rather than something that is created in a static factory method per job and used only once.

The current form seems hard to use right, doesn't actually support multiple concurrent uses, multithreading, etc.

Reply:
changed